RFC: fail `test.data.table()` if non-test code produces warnings

aitap opened the pull request:
Related to #7209. I happened to notice the warning while looking at unrelated failures. Should `test.data.table()` explicitly go looking for warnings produced by code outside the `test()` calls and fail `R CMD check` because of them?

Pros: sometimes there's a good reason to fix the warning (like #7209).

Cons: sometimes warnings happen for stupid reasons outside our direct control. For example, a system that neither currently has a UTF-8 locale set nor has `en_US.UTF-8` installed will cause a warning between tests 2252.2 and 2253.01, because the `setlocale()` call will fail:

https://github.com/Rdatatable/data.table/blob/63339e1bf4a65b1821c0c5a7019270d34e63346e/inst/tests/tests.Rraw#L18619-L18623

aitap committed:
test.data.table: catch warnings outside test code

aitap committed:
Use explicit environment assignment instead of <<-

aitap committed:
Thinko: length -> nrow

MichaelChirico commented:
Personally I like requiring warnings to be handled properly everywhere, I'm a fan of setting `options(warn=2)` in my own {testthat} suites, for example.

If it's really a stupid/spurious warning, just `suppressWarnings()` and move on.

aitap committed:
tests 2253*: suppressWarnings, skip if no UTF-8

aitap committed:
tests: remember tracebacks from warnings

aitap committed:
Thinko: length -> nrow

aitap committed:
Merge branch 'master' into test-warnings

aitap committed:
suppressWarnings(Sys.setlocale(...))

On modern versions of R for Windows, Sys.setlocale() warns for non-UTF-8
locales.

Also comment the previous instance of suppressWarnings().

MichaelChirico commented on R/test.data.table.R:
```diff
@@ -205,7 +205,30 @@ test.data.table = function(script="tests.Rraw", verbose=FALSE, pkg=".", silent=F
+        # nocov start
+        if (!silent && showProgress) print(w)
+        env$warnings = c(env$warnings, list(list(
+          env$prevtest, toString(w),
```
why not `conditionMessage()`? should we give this object a bit more structure, e.g. name the elements for clarity?

MichaelChirico commented on R/test.data.table.R:
```diff
@@ -205,7 +205,30 @@ test.data.table = function(script="tests.Rraw", verbose=FALSE, pkg=".", silent=F
+          env$prevtest, toString(w),
+          paste(
+            vapply_1c(sys.calls(), function(call) {
+              if (length(call) && is.name(call[[1]])) {
```
is `!length(call)` possible?

MichaelChirico commented on inst/tests/tests.Rraw:
```diff
@@ -18615,12 +18615,17 @@ test(2252.2, dt[, let(b=2L)], error = "\\[ was called on a data.table.*not data.
+    on.exit(Sys.setlocale('LC_CTYPE', lc_ctype))
+    TRUE
+  } else FALSE
+}) {
```
it's ugly, but it looks sound.

if anything, maybe we would just move it inside `test.data.table()` like `foreign`:

https://github.com/Rdatatable/data.table/blob/39e2875a17fdf1ef6fe05cf79f48ffd6613808de/R/test.data.table.R#L119-L125

Then we just do, say, `if (supports_utf) local(...)`.

It would also facilitate if this sort of check would make other tests easier, we don't need to duplicate this complicated logic more than once.

aitap committed:
Apply fixes from code review

* conditionMessage() instead of toString()
* apply names earlier
* don't check for zero-length calls in sys.calls()
* be careful not to over-translate an already-translated message

Co-Authored-By: Michael Chirico <michaelchirico4@gmail.com>
